ocaml/parsing/jane_syntax.mli

```diff
@@ -100,6 +100,59 @@ module Immutable_arrays : sig
   val pat_of : loc:Location.t -> pattern -> Parsetree.pattern
 end
 
+module Mode_expr : sig
+  (** [Mode_expr] appears in several places:
+  - let local_ x = ...
+  - local_ exp
+  - local string -> string
+  - {global_ x : int}
+
+  Note that in the first two cases, axes other than locality are not specified;
+  in the second case, other axes are defaulted to legacy. In the last case, we
+  are specifying modalities.
+
+  In the future the three annotations will be quite different, but for now they
+  are all lists of modes/modalities. [Typemexp] has the three different
+  interpretations of the annotation.
+
+  (TODO: in the future we will have mutable(...), which is similar to the second
+  occurrence above and should be covered by this module)
+  *)
+  type t = string Location.loc list Location.loc
+
+  (** The empty mode expression. *)
+  val empty : t
+
+  (** The mode expression containing a single mode constant. *)
+  val singleton : string -> Location.t -> t
+
+  (** The string used to mark attributes/extensions as containing mode
+      expressions.  *)
+  val embedded_name_str : string
```

Comment:
I assume when you say "extension" you mean `Pexp_extension`. What you said is true, but renaming it to `extension_name` also means `attr_of` will be citing `extension_name`, which is strange. Maybe we should have another `attribute_name` which is identical to `extension_name`?

I think that's too much overhead for little gain. I think we should just have a single name that covers all syntatical categories, and it's *coincidental* that it's only used by `Pexp_extension`.

Reply:
I think the coincidence is that the attribute/extension encoding happen to use the same name.

My concrete proposal is:
  * rename the exposed binding to `extension_name`
  * internally to `jane_syntax.ml`, have two unrelated bindings `extension_name` and `attribute_name`. Don't expose the latter. It's what's used in `attribute_of`.

I think this makes the interface of `Jane_syntax` to be more self-documenting.

Reply:
(In other words, I think the approach that you described to have "too much overhead' doesn't in fact have much overhead at all.)

Reply:
Overhead aside - I still don't understand why we should have `attribute_name` and `extension_name` separately - I thought since they belong to the same feature, it would be nice to unify the identification string? Your point is that the exposed string is only used by extensions, which is true. But I don't see how that matters here